lib/src/shared/main/java/com/couchbase/lite/internal/AbstractExecutionService.java

```diff
@@ -75,33 +98,92 @@ public String toString() {
+     * If the executor is stopped with unscheduled pending tasks
+     * (in the pendingTask queue), all of those tasks are simply discarded.
+     * If the pendingTask queue is non-empty, the head task is scheduled.
+     * There is probably no need to dump the pendingTask queue, on failure:
+     * the problem is, almost certainly, tasks that are already scheduled
+     * in the underlying executor.
+     */
     private static class ConcurrentExecutor implements CloseableExecutor {
-        private final Executor executor;
+        @NonNull
+        private final ThreadPoolExecutor executor;
+
+        @GuardedBy("this")
+        @NonNull
+        private final Queue<InstrumentedTask> pendingTasks = new LinkedList<>();
+
+        // a non-null stop latch is the flag that this executor has been stopped
+        @GuardedBy("this")
+        @Nullable
         private CountDownLatch stopLatch;
+
+        @GuardedBy("this")
         private int running;
 
-        private ConcurrentExecutor(Executor executor) { this.executor = executor; }
+        ConcurrentExecutor(@NonNull ThreadPoolExecutor executor) {
+            Preconditions.checkArgNotNull(executor, "executor");
+            this.executor = executor;
+        }
 
+        /**
+         * Schedule a task for concurrent execution.
+         * If there is insufficient room to schedule the task, safely, on the underlying
+         * executor, the task is added to the pendingTask queue and executed when spaces is available.
+         * There are absolutely no guarantees about execution order, on this executor,
+         * particularly once it fails back to using the pending task queue.
+         *
+         * @param task a task for concurrent execution.
+         * @throws ExecutorClosedException    if the executor has been stopped
+         * @throws RejectedExecutionException if the underlying executor rejects the task
+         */
         @Override
-        public synchronized void execute(@NonNull Runnable task) {
-            if (stopLatch != null) { throw new ExecutorClosedExeception("Executor has been stopped"); }
+        public void execute(@NonNull Runnable task) {
+            Preconditions.checkArgNotNull(task, "task");
 
-            try {
-                executor.execute(new InstrumentedTask(task, this::finishTask));
-                running++;
-            }
-            catch (RejectedExecutionException e) {
-                dumpServiceState(executor, e, "size: " + running);
-                throw e;
+            final int n;
+            synchronized (this) {
+                if (stopLatch != null) { throw new ExecutorClosedException("Executor has been stopped"); }
+
+                final InstrumentedTask newTask;
+                if (spaceAvailable()) { newTask = new InstrumentedTask(task, this::finishTask); }
+                else {
+                    newTask = new InstrumentedTask(task, this::scheduleNext);
+                    pendingTasks.add(newTask);
```

Comment:
Part 2 of the comment below.  Since `executeTask` only gets called when there is one or zero items in this queue, it's possibly game over if it reaches 2 and no new tasks will ever execute:

1. This logic is reached, and `scheduleNext` will get executed
2. More tasks come into the `pendingTasks` list from somewhere else
3. `scheduleNext` executes a few, but runs out of space on the executor before finishing, leaving at least 1 item in `pendingTasks`
4. `pendingTasks` receives a second item and no items are ever executed again